src/hotspot/share/opto/output.cpp

```diff
@@ -2022,6 +2022,8 @@ void PhaseOutput::FillExceptionTables(uint cnt, uint *call_returns, uint *inct_s
 
     // Handle implicit null exception table updates
     if (n->is_MachNullCheck()) {
+      assert(n->in(1)->as_Mach()->barrier_data() == 0,
+             "Implicit null checks on memory accesses with barriers are not yet supported");
```

Comment:
I don't see here changes in `lcm.cpp` which would prevent it.

Reply:
I did not make any changes because the current logic in `lcm.cpp` already prevents this, albeit in a rather accidental way: `PhaseCFG::implicit_null_check` requires that all inputs of a candidate memory operation dominate the null check ([here](https://github.com/robcasloz/jdk/blob/d21104ca8ff1eef88a9d87fb78dda3009414b5b8/src/hotspot/share/opto/lcm.cpp#L310-L328)) so that it can be hoisted. This fails if the candidate memory operation has barriers because these always require `MachTemp` nodes, which are placed in the same block as the candidate and break the dominance condition. See a longer explanation [here](https://github.com/openjdk/jdk/pull/19746/files#r1715387255).

Should I add a check to `PhaseCFG::implicit_null_check` to discard these memory accesses more explicitly?

Reply:
> Should I add a check to PhaseCFG::implicit_null_check to discard these memory accesses more explicitly?

Yes, please.

Reply:
Done (commit 71a51bfc).